### PR TITLE
Attempt to fix header collision issue

### LINKED ISSUE #13: Headers are being lowercased
Setting a header in a `z.request` results in the header being lowercased when sent out on the wire.

Example code snippet to reproduce:

```javascript
  const promise = z.request(`https://api.lyft.com/oauth/token`, {
    method: 'POST',
    json: {
      code: bundle.inputData.code,
      grant_type: 'authorization_code'
    },
    headers: {
      'Content-Type': 'application/json',
      'Authorization': auth_header
    }
  });
```

Resulted in this request:

```
POST https://api.lyft.com/oauth/token
user-agent: Zapier
content-type: application/json; charset=utf-8
cache-control: no-cache
authorization: <auth_header>
```

### PR DESCRIPTION
**Pull request by BrunoBernardino**
Fixes #13

### CODE REVIEW

**Review by bcooksey — COMMENTED** [review]
One hesitation with this approach is that the docs for [Object.keys()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/keys) say

> in the same order as that provided by a for...in loop

And then [for...in loop](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...in) says

> iterates over the enumerable properties of an object, in arbitrary order

Can we count on the order provided by `.keys()` to consistently reflect last in? @charettes suggestion about a new object type and storing lower-cased keys but including the casing alongside the value seemed solid, though definitely more work. Thoughts?

**Comment by BrunoBernardino** [comment]
Good point, and indeed, by JS object definition we can't get a guaranteed order on an object's properties/keys, that's why I'm using the same array (so the order is the same in both merging and checking for "duplicates").

That being said, I don't expect this to cause a problem (we're already addressing an edge case that isn't common), but if it _ever_ does, we should probably invest in such a full-fledged solution (that won't be easy or quick at all).

**Comment by xavdid** [comment]
I don't think a full custom object is worth it, but I think we're on the right track here. Instead of adding everything and then removing what looks like duplicates, we do the check when we're modifying the headers. Only add the default key if a case insensitive match doesn't exist already. 

Hopefully we only add defaults in one spot :D

**Comment by xavdid** [comment]
So I've been thinking about this today. 

I think the most important thing is that whatever the user puts in, they get back out. If the API they're using has a case-sensitive header and they enter it correctly only to have us swap the case on them, that's gonna be super frustrating. To that end, it would be most cautious to always check before adding a header via a function like this:

```js
// headers comes in as an object
function addHeader(headers, newKey, newVal) {
  let addKey = true;
  for (let k of Object.keys(headers)) {
    if (k.toLowerCase() === newKey.toLowerCase()) {
      addKey = false;
      break;
    }
  }
  if (addKey) {
    header[newKey] = newVal;
  }
  return headers;
}
```
As long as we use that to add a header anywhere we edit them, we'll be safe. 

That being said, as long as the PR as written definitely won't overwrite user-entered info (and I don't think it will, but i'm a little hazy on when & why `mergeRequests` gets used), then I'm fine with it as is. Liked @BrunoBernardino mentioned, it's a super corner case, we're probably worrying over nothing. 😀

**Review by xavdid — COMMENTED** [review]
So I've been thinking about this today. 

I think the most important thing is that whatever the user puts in, they get back out. If the API they're using has a case-sensitive header and they enter it correctly only to have us swap the case on them, that's gonna be super frustrating. To that end, it would be most cautious to always check before adding a header via a function like this:

```js
// headers comes in as an object
function maybeAddHeader(headers, newKey, newVal) {
  let addKey = true;
  for (let k of Object.keys(headers)) {
    if (k.toLowerCase() === newKey.toLowerCase()) {
      addKey = false;
      break;
    }
  }
  if (addKey) {
    header[newKey] = newVal;
  }
  return headers;
}
```
As long as we use that to add a header anywhere we edit them, we'll be safe. 

That being said, as long as the PR as written definitely won't overwrite user-entered info (and I don't think it will, but i'm a little hazy on when & why `mergeRequests` gets used), then I'm fine with it as is. Liked @BrunoBernardino mentioned, it's a super corner case, we're probably worrying over nothing. 😀

**Comment by BrunoBernardino** [comment]
I agree and love the suggestion @xavdid — like you mentioned, with this PR as is, we should be covered. Do note that to apply your suggestion would make it so that in a merge, we keep the old headers instead of the new (maybe that's a better behavior? Not sure), and we'd have to change how the requests get merged, so we check for each header before being merged, rather than merging and then checking if there's any duplicates (case-insensitive-wise). I think that'd result in more iterations over the keys — I might be looking at it wrongly, though.

I also think this PR as is keeps the surprises to a minimum, meaning that if I add `{'api-key': 'ab'}` and then `{'Api-Key': 'cd'}`, I'll get `{'Api-Key': 'cd'}`, but with your suggestion I think I'd get `{'api-key': 'ab'}`. Maybe that's fine, but it's a different behavior than what we currently have (same key gets latest value), and we would need to change for any merging, not just duplicate-finding.

**Comment by BrunoBernardino** [comment]
So, are we good to merge @xavdid / @bcooksey ? Or should we add more test cases?

**Comment by BrunoBernardino** [comment]
Alright, after some brainstorming I'll try to do a better merge and loop through the keys less, trying to guarantee the order of the keys, per request, are kept, potentially applying the clean and deleting of keys while doing our own-flavoured merge.

File: src/tools/request-merge.js [hunk 1]
```diff
@@ -4,6 +4,51 @@ const _ = require('lodash');
...
 
+// Do a merge with case-insensitive keys, and drop empty header keys
+const caseInsensitiveMerge = (requestOne, requestTwo, requestThree) => {
+  // This creates copies/clones
+  requestOne = requestClean(requestOne);
+  requestTwo = requestClean(requestTwo);
+  requestThree = requestClean(requestThree);
+
+  // This is a very quick & efficient merge for all of request's properties
+  const mergedRequest = _.merge.apply(_, [requestOne, requestTwo, requestThree]);
```
Comment (ibolmo) [inline src/tools/request-merge.js]: You can do: `_.merge(requestOne, requestTwo, requestThree)`

File: src/tools/request-merge.js [hunk 2]
```diff
@@ -4,6 +4,51 @@ const _ = require('lodash');
...
+  const requestTwoHeaders = requestTwo.headers || {};
+  const requestOneHeaders = requestOne.headers || {};
+
+  // Check, in order, which keys to add (if they're not duplicate)
+  [requestTwoHeaders, requestOneHeaders].forEach((requestHeaders) => {
+    const existingKeys = Object.keys(mergedRequestHeaders);
+    const requestKeys = Object.keys(requestHeaders);
+
+    // We will loop through every header, and if we find no (case-insensitive) match, we'll add it to mergedRequest
+    requestKeys.forEach((checkingKey) => {
```
Comment (ibolmo) [inline src/tools/request-merge.js]: Could use a lowerCase hash and just check if one is already found? E.g.

```js
// warning pseudo
seen = {'authorization': ....}
lower = key.toLowerCase();
seen[lower] = seen[lower] || requestHeaders[key];
```

File: src/tools/request-merge.js [hunk 3]
```diff
@@ -4,6 +4,51 @@ const _ = require('lodash');
...
+      // Only add if the key doesn't exist yet
+      if (foundKeyIndex === -1) {
+        mergedRequestHeaders[checkingKey] = requestHeaders[checkingKey];
+      }
+    });
+  });
+
+  // Remove any keys with DROP_DIRECTIVE, after all merging happened
+  Object.keys(mergedRequestHeaders).forEach((key) => {
+    if (mergedRequestHeaders[key] === requestClean.DROP_DIRECTIVE) {
```
Comment (ibolmo) [inline src/tools/request-merge.js]: I wonder if this could be checked inline while doing the looping above.

**Review by ibolmo — COMMENTED** [review]
Didn't mean to start a review. Now that I have a bit more time, wanted to help :D

File: src/tools/request-merge.js (hunk 3 above)
Comment (BrunoBernardino) [reply src/tools/request-merge.js]: Great question! I tried, and the problem is the `DROP_DIRECTIVE` can be in last request's headers, which means that if I delete it then, the other requests can add it back. We have to wait until all the merging is finished.

File: src/tools/request-merge.js (hunk 2 above)
Comment (BrunoBernardino) [reply src/tools/request-merge.js]: Great point. Do you think it'll improve readability or performance considerably? I'm thinking the header object will never have _many_ keys, but I might be missing something.

Comment (ibolmo) [reply src/tools/request-merge.js]: Up to the team. Personally the hash lookup is very clear.

File: src/tools/request-merge.js [hunk 4]
```diff
@@ -4,6 +4,51 @@ const _ = require('lodash');
...
+  // This creates copies/clones
+  requestOne = requestClean(requestOne);
+  requestTwo = requestClean(requestTwo);
+  requestThree = requestClean(requestThree);
+
+  // This is a very quick & efficient merge for all of request's properties
+  const mergedRequest = _.merge(requestOne, requestTwo, requestThree);
+
+  // Now to cleanup headers, we start on the last request (the one with priority) and work backwards to add the keys that don't already exist
+  // NOTE: This is done "manually" instead of a _.merge or Object.assign() because we need case-insensitivity
```
Comment (bcooksey) [inline src/tools/request-merge.js:18]: This "why" type of note is great. Some of the other comments in this function are more "what" type comments that should be ok to nuke.

File: src/tools/request-merge.js (hunk 2 above)
Comment (bcooksey) [reply src/tools/request-merge.js]: @ibolmo See what you are saying. The downside there is that you then loose the original casing, so you'd have to track `seen` plus an object that has the casing we want, or else track the desired key along with the value in the insensitive version. At that point, both seem about equal.

**Review by bcooksey — APPROVED** [review]

**Comment by bryanhelmig** [comment]
I'm not 100% up to speed on our current plan of attack but alternate technique - now that Lambda supports Node 6.10 we could use `Proxy` and a `set` trap to provide a nativish interface there.

Also, we have prior art (which we use via `fetch` we could lean on) https://github.com/bitinn/node-fetch/blob/master/src/headers.js.

**Pull request merged**